### ISSUE TITLE
StdinReader stops working and CPU usage is high

### ISSUE THREAD

**Issue by voidcontext** [jaor/xmobar#417]
Hello,

After some time xmobar stops displaying the output of xmonad while CPU usage is high (100% on a single core by the xmobar process). Unfortunately couldn't figure out what is triggering the issue, any idea how to get to the bottom of this? If remove `StdinReader` from the commands and from the output template then everything works fine.

My xmobarrc:

```haskell
Config { font = "xft:Monaco:size=8:antialias=true"
       , additionalFonts = []
       , borderColor = "black"
       , border = TopB
       , bgColor = "black"
       , fgColor = "grey"
       , alpha = 255
       , position = Top
       , textOffset = -1
       , iconOffset = -1
       , lowerOnStart = True
       , pickBroadest = False
       , persistent = False
       , hideOnStart = False
       , iconRoot = "."
       , allDesktops = True
       , overrideRedirect = False
       , commands = [ Run Network "eth0" ["-L","0","-H","32",
                                          "--normal","green","--high","red"] 10
                    , Run Network "wlp3s0" ["-L","0","-H","32",
                                          "--normal","green","--high","red"] 10
                    , Run Wireless "wlp3s0" ["-L","0","-H","32",
                                          "--normal","green","--high","red"] 10
                    , Run Cpu ["-L","3","-H","50",
                               "--normal","green","--high","red"] 10
                    , Run Memory ["-t","Mem: <usedratio>%"] 10
                    , Run Swap [] 10
                    , Run Date "%a %b %_d %Y %H:%M:%S" "date" 10
                    , Run StdinReader
                    ]
       , sepChar = "%"
       , alignSep = "}{"
       , template = "%StdinReader% | %cpu% | %memory% * %swap% | %eth0% - %wlp3s0% %wlp3s0wi% }\
                   \{ <fc=#ee9a00>%date%</fc>"
       }
```

and my xmonad config:

```haskell
import XMonad
import XMonad.Hooks.DynamicLog
import XMonad.Util.EZConfig

-- The main function.
main = xmonad =<< statusBar myBar myPP toggleStrutsKey myConfig

-- Command to launch the bar.
myBar = "xmobar"

-- Custom PP, configure it as you like. It determines what is being written to the bar.
myPP = xmobarPP -- { ppCurrent = xmobarColor "#429942" "" . wrap "<" ">" }

-- Key binding to toggle the gap for the bar.
toggleStrutsKey XConfig {XMonad.modMask = modMask} = (modMask, xK_b)

-- Main configuration, override the defaults to your liking.
myConfig = defaultConfig
  { terminal = "$HOME/.nix-profile/bin/gnome-terminal"
  }
  `removeKeysP` [ "M-w"
                , "M-e"
                , "M-r"
                ]
```

**Comment by Xitian9** [comment]
I'm just chiming in to add that I'm experiencing the same problem. This is happening with both versions 0.32 and 0.30. I was previously using xmobar-0.30 for a long time without encountering this, but it appeared after I installed from nixpkgs-unstable (I had previously been using portage from Gentoo).

**Comment by Xitian9** [comment]
The problem occurs with the following minimal xmobarrc:

```
Config { allDesktops = True
       , overrideRedirect = False
       , borderColor = "black"
       , border = TopB
       , bgColor = "black"
       , fgColor = "grey"
       , position = Top
       , lowerOnStart = True
       , persistent = True
       , hideOnStart = False
       , textOffset = 16
       , iconOffset = 10
       , iconRoot = "/home/stephen/.local/share/icons/dzen"
       , template =
           "%StdinReader%"
       , commands =
            [ Run StdinReader
            ]
       }
```

Since I see that @voidcontext seems to also be using nix, perhaps the problem is with one of the libraries in our respective package sets.

Let me know what other information might be helpful.

**Comment by Xitian9** [comment]
Aha! The problem seems to occur the instant StdinReader tries to dispay a non-ascii character. This might be related to simonmichael/hledger#1033.

**Comment by Xitian9** [comment]
Yes, the problem seems to be fixed by putting the following declaration somewhere xmobar can see it, e.g. .xsession, .zprofile, .bash_profile.

```
export LOCALE_ARCHIVE="$(nix-build --no-out-link '<nixpkgs>' -A glibcLocales)/lib/locale/locale-archive"
```

**Comment by voidcontext** [comment]
@Xitian9 nice catch, I'll try this our later and will report back the results. Thanks for tracking this down!

**Comment by voidcontext** [comment]
One way to reproduce the issue: login through `gdm` into `xmonad`, then start `gnome-terminal` and open "Edit" -> "Preferences".

@Xitian9 the fix you suggested works, but on Ubuntu (19.10) the export needs to be in the `.xsessionrc` file otherwise it won't be loaded.

Thanks again!
